airbyte_cdk/manifest_server/routers/manifest.py

```diff
@@ -0,0 +1,158 @@
+) -> ConcurrentDeclarativeSource[Optional[List[AirbyteStateMessage]]]:
+    """Wrapper around build_source that converts ValidationError to HTTPException."""
+    try:
+        return build_source(
+            manifest_dict,
+            catalog,
+            config_dict,
+            state,
+            record_limit,
+            page_limit,
+            slice_limit,
+        )
+    except jsonschema.exceptions.ValidationError as e:
+        raise HTTPException(status_code=400, detail=f"Invalid manifest: {e.message}")
+
+
+router = APIRouter(
+    prefix="/manifest",
+    tags=["manifest"],
+    dependencies=[Depends(verify_jwt_token)],
+)
+
+
+@router.post("/test_read", operation_id="testRead")
+def test_read(request: StreamTestReadRequest) -> StreamRead:
+    """
+    Test reading from a specific stream in the manifest.
+    """
+    config_dict = request.config.model_dump()
+
+    catalog = build_catalog(request.stream_name)
+    converted_state = [AirbyteStateMessageSerializer.load(state) for state in request.state]
+
+    if request.custom_components_code:
+        config_dict[INJECTED_COMPONENTS_PY] = request.custom_components_code
+        config_dict[INJECTED_COMPONENTS_PY_CHECKSUMS] = {
+            "md5": hashlib.md5(request.custom_components_code.encode()).hexdigest()
+        }
+
+    source = safe_build_source(
+        request.manifest.model_dump(),
+        config_dict,
+        catalog,
+        converted_state,
+        request.page_limit,
+        request.slice_limit,
+        request.record_limit,
+    )
+
+    runner = ManifestCommandProcessor(source)
+    cdk_result = runner.test_read(
+        config_dict,
+        catalog,
+        converted_state,
+        request.record_limit,
+        request.page_limit,
+        request.slice_limit,
+    )
+    return StreamRead.model_validate(asdict(cdk_result))
+
+
+@router.post("/check", operation_id="check")
+def check(request: CheckRequest) -> CheckResponse:
+    """Check configuration against a manifest"""
+    source = safe_build_source(request.manifest.model_dump(), request.config.model_dump())
+    runner = ManifestCommandProcessor(source)
+    success, message = runner.check_connection(request.config.model_dump())
+    return CheckResponse(success=success, message=message)
+
+
+@router.post("/discover", operation_id="discover")
+def discover(request: DiscoverRequest) -> DiscoverResponse:
+    """Discover streams from a manifest"""
+    source = safe_build_source(request.manifest.model_dump(), request.config.model_dump())
+    runner = ManifestCommandProcessor(source)
+    catalog = runner.discover(request.config.model_dump())
+    if catalog is None:
+        raise HTTPException(status_code=422, detail="Connector did not return a discovered catalog")
+    return DiscoverResponse(catalog=catalog)
+
+
+@router.post("/resolve", operation_id="resolve")
+def resolve(request: ResolveRequest) -> ManifestResponse:
+    """Resolve a manifest to its final configuration."""
```

Comment:
Nit: I understand "full resolve" expands dynamic schemas, but I still don't fully understand what this step does, even after working with it for a while.

What do you think of adding to the docstring an example of what gets resolved in this step and when this should be called? Also could be helpful to explain if the resolved version of the manifest replaces the original, or is it just used for json schema validation or other preprocessing needed for json schema validation, and/or something else? Thanks!

Reply:
To be honest I also don't fully have my head around when this is necessary lol. I added to the docstring with the explanation Lake gave me. It seems like this is only called from the builder in a specific case.